### PR TITLE
Update dotnet sdk to support dotnet cli F# and VS F#

### PR DESCRIPTION
**Pull request by KevinRansom**
Update the sdk to add a couple of shims for a props file and targets file deployed with the F# compiler.

This adds 2 shims that redirect to the fsharp props and targets file when invoked via the sdk.

On desktop msbuild, I.e on windows with the desktop and under vs, uses VS installed in the normal locations.

On dotnet cli uses targets from the coreclr compiler nuget package.

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (dasMulli) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: What about running on Mono? On mono, the runtime type isn't core and and there aren't VS-specific targets.

**Comment by KevinRansom** [comment]
I'm guessing the list dnfclas uses needs updating.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (KevinRansom) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: I will chat with the sdk team tomorrow to find out about what happens with the sdk on mono.

**Comment by KevinRansom** [comment]
@dotnet-bot test this please.

**Comment by enricosada** [comment]
@KevinRansom why these are needed? 
how is `Microsoft.NET.Sdk.FSharp` different than `FSharp.NET.Sdk`? 
with

```
  <LanguageTargets Condition="'$(MSBuildProjectExtension)' == '.fsproj'">$(MSBuildToolsPath)\Microsoft.FSharp.targets</LanguageTargets>

```

you are pratically overriding `FSharp.NET.Sdk`.

Cannot we just bundle the f# target in `FSharp.NET.Sdk` or `FSharp.Compiler.Tools` packages? So exist only one target for all msbuild host runtimes (mono/vs/dotnetcore)

i dont understand why not, we can also redirect to bundled with VS for f# one there if needed.

**Comment by KevinRansom** [comment]
@enricosada.

The dotnet cli will deploy C#, VS and F# inbox.  This arrangement will allow VS, Desktop MSBuild and DotNet Cli to build FSharp apps.  And will ensure that CPS will not have to do a project restore to get the SDK and compiler on project load.

However I expect existing F# SDK projects to still work fine, given that you override LanguageTargets.

**Comment by enricosada** [comment]
As a note.

From [FSharp.NET.Sdk v1.0.3](https://github.com/dotnet/netcorecli-fsc/releases/tag/v1.0.3), the new fsproj works in mono 5, msbuild .net full (VS 2017), and .net core OOTB.

- on mono 5, when run with `msbuild`, it will run `mono fsc.exe` (net40)  from `FSharp.Compiler.Tools` package (finalized in https://bugzilla.xamarin.com/show_bug.cgi?id=55626 with mono `5.0.0.89`) 
- on msbuild .net full, (like VS), will run `fsc.exe` (net40) from `FSharp.Compiler.Tools` package.
- on `dotnet msbuild` (dotnet/cli) will wun `dotnet fsc.dll` (netcoreapp1.0 atm) from from `FSharp.Compiler.Tools` package.

and works when fsproj target  `net4*`, `netcoreapp*`, `netstandard*`, etc, without changes.

It also already import what's needed for Roslyn project system with [FSharp.NET.Sdk v1.0.4](https://github.com/dotnet/netcorecli-fsc/releases/tag/v1.0.4)

**Comment by enricosada** [comment]
@KevinRansom both already fixed.

> This arrangement will allow VS, Desktop MSBuild and DotNet Cli to build FSharp apps

It already does as 1.0.3, see release notes. or try locally.

> And will ensure that CPS will not have to do a project restore to get the SDK and compiler on project load.

It doesnt need to, `1.0.4-bundled` already works for Roslyn ps, see https://github.com/dotnet/project-system/issues/1857#issuecomment-298266904
And is the same package embedded in all msbuild hosts and sdks.

**Comment by enricosada** [comment]
@KevinRansom can we discuss in skype/slack chat a bit about that?
i dont understand what's the issue here, or if i miss something about where sdk is going for VS (i think i had a clear vision before, but maybe i am missing something big)

**Comment by dsyme** [comment]
I haven’t thought much about this but I think I vaguely understand what’s going on.

Shipping F# support in-the-box “just like C# and VB” would be great from many perspectives. I hadn’t realized it was even on the cards.  

Can I check which if @enricosada's criteria would be fulfilled by this? My guess is as follows: 

Yes:
* [x] target and msbuild tasks from package
* [x] everything works on mono/msbuild.exe desktop/dotnet cli. (Assuming the F# targets for .NET Framework work on Mono)
* [x] msbuild host runs the "natural" fsc (mono/desktio -> fsc net40, cli -> netcoreapp1.0) with v1.0.3

No:
* [ ] fsc from package, upgradable, just change fsharp.compiler.tools version in fsproj. NO: now opt-in by modifying project file to use different targets?? Or Set FscPath? Or??

I don't know:
* [ ] works with roslyn project system (v1.0.4).
* [ ] templates are ok.
* [ ] works with sdk 1.0 and 2.0.
* [ ] all targets (publish/pack/etc) works, crossgen (multi framework) works. on mono too.
* [ ] support choose fsc based on bundled runtime https://github.com/fsharp/fsharp/issues/737 so with sdk 2.0 with use fsc netcoreapp2.0
* [ ] fsc can be overridden with old FscToolPath/FscToolExe

**Comment by dasMulli** [comment]
I'd be interested to know why this needs to be done in Microsoft.NET.Sdk and not in FSharp.NET.Sdk, since fsproj templates generate the projects using `<Project Sdk="FSharp.NET.Sdk;Microsoft.NET.Sdk">`.

**Comment by enricosada** [comment]
Just one note but important. 

All these features doesn the matter @dsyme. It was the incremental process ihmo to improve. 

You can do exactly the same bundling both current Sdk directory and content of fsharp.net.sdk. 
And remove fsharp.net.sdk and jmpory these directly in Ms.net.sdk
Wil work exactly the same as today. 

I have designed and implemented like that because like that I can upgrade the target files with a package push because:

- was not done completly (I fixed vs/FSAC intellisense after 1.0, etc)
- once shipped the cli/vs I cannot update it to fix bugs or enhance.  
- tenplate are fixed ( lot of pain in previous preview about that)
- remote sdk (from nupkg) are planned afaik

Like that I can fix and add features, as done with multiple releases https://github.com/dotnet/netcorecli-fsc/releases already. 

That's why. Is separate packaging a bit for a reason. But integrated completly in the sdk.

But probably doesn't matter anymore. Or Maybe yes. Dont know, but I need to stop there a bit, until I understand what is this pr adding. Or not. 
Sry just a bit disappointed but maybe I am misunderstanding.

**Comment by dsplaisted** [comment]
@dotnet-bot test Windows_NT_FullFramework Debug

**Comment by dsyme** [comment]
I've been thinking through the pros and cons of this, where "this" is having a default install of the .NET Core SDK include a default version of the F# .NET Core support.

First, no matter what, this would be a huge success for the F# work by @enricosada and other FSharp.NET SDK contributors have done, which has made it possible to take steps towards having F# support as part of the default install of .NET Core SDK.  That's no small achievement.

Here are the advantages I can see for to having a default version of F# support available as part of a default install
* We get to NGEN or crossgen the F# compiler on install.  This is a considerable thing
* Only one copy of the F# compiler gets installed when using default settings
* A new F# project + build will not require any package references to be downloaded.  
* The default project files can be a little simpler (no reference to the compiler package)
* Anything that gets done for shared targets should just start to work for F#

All in all 1st class support in the default install aligned with the way C# does this stuff is a good thing to aim for.  

However, that said, we should also work on making sure the opt-in scenarios (e.g. new F# tools in a package) work really well.

**Comment by dsyme** [comment]
@KevinRansom Have you looked at the FSharp.NET SDK bugs list like [this one](https://github.com/dotnet/netcorecli-fsc/issues/93)? The list is an interesting guide. It looks like we need to add F# support to [this code](https://github.com/Microsoft/msbuild/blob/master/src/Tasks/WriteCodeFragment.cs#L294). I think this would need to be done regardless of how the FSharp.NET SDK functionality is gradually integrated?

**Comment by KevinRansom** [comment]
@dsyme.

yes it looks like we may need to fix that.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: Actually, I think we should have a looser coupling where there's a way to signal "don't use the FSharp support bundled with the .NET Core Sdk. 
``` 
<PropertyGroup>
  <!-- For compatibility with older versions of FSharp.NET.Sdk that do not set UseBuiltinFSharp. -->
  <UseBundledFSharpSupport Condition="'$(UseBundledFSharpSupport)' and $(_FscTaskAssemblyPath_net)' != ">false</UseBundledFSharpTargets>

  <UseBundedFSharpSupport Condition="'$(UseBundledFSharpSupport)' == ''">true</UseBundledFSharpSupport>
```

So the contract for someone wanting to replace the FSharp toolkit is to set UseBundledFSharpSupport and we only use the check for a known FSharp.NET.Sdk variable to avoid breaking existing versions.

**Comment by enricosada** [comment]
I am leaving the last unrequested feedback, because i see my previous comments and questions (github and mail) didnt got an answer (or just a "i am doing it because is needed, but not explaining why"), so that mean, is doesnt matter.

**I think this is going in the wrong direction**

It's not about offline scenario, or ngen, or use bundled fsc. both can be done in a different way, easier.

**Ihmo main question is if we want to use just one source for adding only the targets/tasks shared between all msbuild (mono/cli/VS) for f#, to expect same behaviour**
Current `FSharp.NET.Sdk` does that, with plans about how to fix the other real issues.

**Here is pratically forking in two, VS and others**. 

This is wrong, and not needed. If VS need something (like [had with roslyn ps](https://github.com/dotnet/netcorecli-fsc/blob/26db12d7f3b50438a91042b95fab91acede1b4b1/src/FSharp.NET.Sdk/build/FSharp.NET.Current.Sdk.targets#L93-L98) ), can be done in a shared way.

But is not this PR.

Also is not going to matter, because to use this, the default VS template will not add the `FSharp.NET.Sdk` packages, so there is going to be a split anyway. 
and that mean all templates (cli) will follow, not adding that in the future -> idea of shared implementation (`FSharp.NET.Sdk`) useless, or crippled.

So just remove `FSharp.NET.Sdk` now. Dont add a extensibility, who is not going to be used, and will break in the future, complicating things. Bundle the same with `dotnet/cli`.

And i am not saying that because i worked on the `FSharp.NET.Sdk`, but because i

**Comment by nguerrera** [comment]
> fsc from package, upgradable, just change fsharp.compiler.tools version in fsproj. NO: now opt-in by modifying project file to use different targets?? Or Set FscPath? Or??

I think the ideal experience for this would be to add a PackageReference that carries appropriate targets, tasks, and tools to supply its own fsc. IIRC there is/was a csc/vbc nupkg for updating the compiler out-of-band though I don't know if it's been made to work with core projects. cc @jaredpar

And ideally, the CLI bundling would pull bits from the same tooling nupkg that can also be referenced at a later version to get a newer compiler.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (KevinRansom) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: @nguerrera 
I will want existing fsharp.sdk projects to work.  This is a way of detecting that it is an existing project using the existing FSharp.Sdk.  new projects should use the dotnet sdk and so won't set these variables.  Also the fsharp sdk can go on evolving and add a clearer signal, I would like that, however, existing projects still need to build.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (KevinRansom) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: This file is the existing VS shim to the f# compiler.   The Microsoft.NET.Sdk.FSharp.props that ships with the SDK is the SDK shim that switches between dotnet cli deployed compiler and the desktop compiler dependent on which msbuild is running.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (KevinRansom) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: This is to switch between the desktop version of the compiler or the coreclr version, depending on whether the coreclr version of msbuild is running or not.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: Notice that I still have a check for the variables for compat with existing projects. The idea is that we expose a "public" contract for overriding the toolset and grandfather in exsiting versions of FSharp.NET.Sdk for existing projects.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (KevinRansom) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: Yeah comment is wrong.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: I'm surprised that there is an existing VS file with the name Microsoft.NET.Sdk.FSharp.props.

**Comment by KevinRansom** [comment]
@nguerrera @enricosada I expect we can fairly easily deploy a new F# compiler to the dotnet cli --- it's just a dotnet publish after all.  We may need to add a property and update the shims to switch between a preview and the newly published compiler.

**Comment by dsyme** [comment]
I have created an F# RFC to allow us to be fully transparent about the proposal, its motivations and its ramifications

	https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1032-fsharp-in-dotnet-sdk.md

Could all people on this thread please review the content of the RFC and adjust it as necessary via PR or direct edit.  If there is anything we are not being transparent about, please add it,

The main thing is that we all get on the same page, have a record of the motivations for and details of the work, and that we all work together to get the best experience for F#

**Review by dsyme — COMMENTED** [review]
Please very carefully review Microsoft.NET.Sdk.FSharp.targets against the corresponding file in FSharp.NET.Sdk as mentioned above.  Likewise review the whole contents of FSharp.NET.Sdk and ensure that all functionality there is implemented here.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (dsyme) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: @nguerrera  The file is being added here: https://github.com/Microsoft/visualfsharp/pull/2993/

**Comment by KevinRansom** [comment]
@nguerrera 

So I simplified it, now we just add two shims which take care of sdk details and then delegate to files deployed with the F# compiler.
The F# files will handle specific details of dealing with the compiler and f# projects.

The shims will detect running under FSharp.Sdk and if so do essentially nothing.  If not running under the FSharp.Sdk, it takes care of some stuff necessary to make TargetFrameworks work, and a few other necessary details.

**Comment by KevinRansom** [comment]
@nguerrera 
To avoid the naming confusion of the targets files between repos.  I renamed the shims to:
Microsoft.FSharp.NetSdk.targets
Microsoft.FSharp.NetSdk.props

My head no longer explodes when trying to explain or think about them.

This reads the Microsoft FSharp NetSdk targets and props files.

**Comment by 7sharp9** [comment]
@dsyme 
>We get to NGEN or crossgen the F# compiler on install. This is a considerable thing - and absolutely crucial for regular compilation of large projects on Windows

So this works on non Windows systems?

Reading through here this seems to add an even more fragile process with respect to non Windows systems, I think @enricosada approach was rapid to patch and keep up to date with mono issues etc.

**Comment by cartermp** [comment]
@7sharp9 Let's direct any conversation [here](https://github.com/fsharp/fslang-design/issues/188), but I'll answer your immediate questions about this PR since they're more technical in nature:

> So this works on non Windows systems?

Yes, since this is for supporting .NET Core and .NET Standard.

> Reading through here this seems to add an even more fragile process with respect to non Windows systems, I think @enricosada approach was rapid to patch and keep up to date with mono issues etc.

Enrico's package was also for .NET Core and .NET Standard, except it was also out of band.  So while it could react to changes faster in theory, in practice @enricosada would constantly have to chase breaking changes from the SDK.  In-band support for F# in the SDK (this PR) makes a breaking change to F# a breaking change for the SDK.  NGEN for Windows pieces doesn't affect the other targets; that is, we're not pushing Windows-isms into this which make other OSes "less" supported.  Doing so would not only be harmful, but against the goals of .NET Core in the first place.

**Comment by dsyme** [comment]
> Reading through here this seems to add an even more fragile process with respect to non Windows systems, ...

I don't think this is specifically the case.  But could you give an example please so we can best solve the problem? Either here or in [the discussion thread of the RFC](https://github.com/fsharp/fslang-design/issues/188).

File: src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.targets
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.targets]: Why the difference in path here?

Comment (KevinRansom) [inline src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.targets]: The path needs to be different, because Microsoft.FSharp.targets does not ship with msbuild.

msbuild has always used:  $(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\FSharp to find target files.

This PR:
adds a target and a props file in the sdk\build directory, that points msbuild to the correct location.

**Comment by dsyme** [comment]
@KevinRansom Great progress.  What about testing for resources?  That's been a major pain point for the F# build targets historically

**Comment by KevinRansom** [comment]
@dsyme 

I was in the process of adding support for the portable profiles.  However, the dotnet sdk doesn't support them natively at this time.  I also noticed that VS no longer ships C# and VB support for creating new portable profiles.

I believe the notion is that there is a standard library that matches various portable profiles and that the long term approach is to use those.  Anyway, that would change the way we have to build our various fsharp.core libraries ... work for another day...

No the .bak shouldn't be there ..

As for importing the targets twice the intent is to set a variable with the name of the path to import and then import that once.  The variable should only be set if it hasn't already been set.  When I find out how to discover the F# compiler in mono, there may be a third probed location.  Unless I can convince them to deploy the compiler in the fsharp subdirectory of msbuild binaries.

Kevin

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.targets]: This is incorrect. We should not be forcing how common targets import their extensions.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.targets]: This isn't right either. In cross-targeting case, anything serving as LanguageTargets needs to import Microsoft.Common.CrossTargeting.targets *instead of* Microsoft.Common.targets (not in addition to it). When we're in the "outer build", there are supposed to be very few targets: (basically just build, rebuild, clean) that know how to loop over each target framework and call the "inner build", where the usual cast of targets (CoreCompile and friends) can continue to deal with just one target framework. This is overriding the inner build targets with the outer build targets, but leaving a bunch of inner build targets dangling that would do incorrect things if they were ever invoked.

In the cross-targeting case, this is all that VB and C# language targets do:

https://github.com/Microsoft/msbuild/blob/master/src/Tasks/Microsoft.CSharp.CrossTargeting.targets
https://github.com/Microsoft/msbuild/blob/master/src/Tasks/Microsoft.VisualBasic.CrossTargeting.targets

**Comment by KevinRansom** [comment]
@nguerrera Done, will work on a PR for testing.  But I kind of need this in, and the dotnet cli in, before I can enable the tests.

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.targets]: This comment is repeated below where it actually applies. This copy can be removed.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: On both lines, `'$(FSharpTargetsShim)'  == ''` should be `'$(FSharpPropsShim)' == ''`
On the second line, we should be checking for existence of props not targets.

Comment (KevinRansom) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: thanks nick.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: As discussed offline, I still want this to be decoupled such that everywhere else we're not referring to BuildingUnderFSharpSdk but more generally, we're shutting off the built-in fsharp targets.

**Review by nguerrera — APPROVED** [review]
Please squash this before merging.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.FSharp.props]: nit: duplicated comment

**Comment by nguerrera** [comment]
@MattGertz for preview2 approval

**Comment by MattGertz** [comment]
Approved pending test success.

**Comment by nguerrera** [comment]
@dotnet-bot Test Windows_NT_FullFramework Release

**Comment by KevinRansom** [comment]
Thank you.

**Pull request merged**
